pkg/serviceaccounttoken/secret.go

```diff
@@ -79,21 +63,38 @@ func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrol
 
 	secret, err := ServiceAccountSecret(ctx, sa, secretLister, secretClient)
 	if err != nil {
-		return nil, fmt.Errorf("error looking up secret for service account [%s:%s]: %w", sa.Namespace, sa.Name, err)
+		return nil, fmt.Errorf("error looking up secret for service account %s: %w", logKeyFromObject(sa), err)
 	}
 
 	if secret == nil {
-		secret, err = createServiceAccountSecret(ctx, sa, secretLister, secretClient, lockKey)
+		secret, err = createServiceAccountSecret(ctx, sa, secretLister, secretClient)
+		if err != nil {
+			return nil, err
+		}
+
+		sa, updated, err := annotateSAWithSecret(ctx, sa, secret, clientSet.CoreV1().ServiceAccounts(sa.Namespace), secretClient)
```

Comment:
I think there's still room for duplicates in this part of the code. The situation I'm thinking about is the following:
- Two Rancher replicas (or threads)
- Replica 1 receives a request, and updates the service account
- Replica 2 receives a request. It gets the new updated service account with the annotation.
- Since Replica 2 is using a cache, when it lists secrets through "ServiceAccountSecret", it doesn't see the new secret yet.
- Replica 2 creates a new secret (so there's now 2 secrets)
- `annotateSAWithSecret` sees that the annotation is on the service account, so it doesn't update the annotation, or delete the current secret.
- We use the new secret that we just made rather than the original one

Now future requests will eventually clean this up since ServiceAccountSecret will try to delete "duplicate" secrets. However, I think this could happen in ways that we don't prefer (such as by deleting a token before a thread can use it to make a call, returning an error to the user, or by deleting the secret listed on the service account, leaving the actual token that we are using different than what is listed on the secret).

High level, I think that this can be fixed by adjusting "ServiceAccountSecret" to take this new approach into account. More specifically, maybe we can have this read the annotation and extract the specific secret rather than doing a list and taking the first secret that belongs to this ServiceAccount. We'd still need to account for cleanup here of the existing duplicates, but maybe that's better done in a standalone fashion rather than evaluated on every request (or only evaluated when we add the annotation for the first time).